File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/Row.cs
using SeatsSuggestions.DeepModeling.AdjacentSeatingPlace;
using SeatsSuggestions.DeepModeling.MiddleOfTheRow;
using Value;

namespace SeatsSuggestions;

public class Row(string name, List<SeatingPlace> seatingPlaces) : ValueType<Row>
{
    private string Name { get; } = name;
    public List<SeatingPlace> SeatingPlaces { get; } = seatingPlaces;

    public SeatingOptionIsSuggested SuggestSeatingOption(int partyRequested, PricingCategory pricingCategory)
    { 
        var seatAllocation = new SeatingOptionIsSuggested(partyRequested, pricingCategory);
        
        foreach (var seat in OfferAdjacentSeatsNearerTheMiddleOfRow(partyRequested, pricingCategory))
        {
            if (seat.IsAvailable() && seat.MatchCategory(pricingCategory))
            {
                seatAllocation.AddSeat(seat);
                if (seatAllocation.MatchExpectation())
                {
                    return seatAllocation;
                }
            }
        }
        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public Row AddSeatingPlace(SeatingPlace seatingPlace)
    {
        return new Row(Name, [..SeatingPlaces, seatingPlace]);
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return [Name, new ListByValue<SeatingPlace>(SeatingPlaces)];
    }

    public Row Allocate(SeatingPlace seatingPlacesSuggested)
    {
        var seatingPlaces = SeatingPlaces.Select(seatingPlace => seatingPlacesSuggested.IsSameLocation(seatingPlace)
                ? seatingPlacesSuggested.Allocate()
                : seatingPlace)
            .ToList();
        
        return new Row(Name, seatingPlaces);
    }
    
    private IEnumerable<SeatingPlace> OfferAdjacentSeatsNearerTheMiddleOfRow(int partyRequested, PricingCategory pricingCategory)
    {
        var offerSeatsNearerTheMiddleOfTheRow = TheMiddleOfTheRow.OfferSeatsNearerTheMiddleOfTheRow(this);

        if (partyRequested > 1)
        {
            var offerAdjacentSeatsNearerTheMiddleOfRow = AdjacentSeatingPlaces
                .OfferAdjacentSeatingPlace(offerSeatsNearerTheMiddleOfTheRow, pricingCategory, partyRequested);
            
            return offerAdjacentSeatsNearerTheMiddleOfRow;
        }

        return offerSeatsNearerTheMiddleOfTheRow
            .OrderBy(s => s.DistanceFromTheMiddleOfTheRow)
            .Select(s => s.SeatingPlace)
            .Where(s => s.MatchCategory(pricingCategory))
            .Where(s => s.IsAvailable())
            .Take(partyRequested)
            .ToList();
    }
}

File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestionsDomain/src/main/java/org/octo/seatingplacessuggestions/domain/seatingplacesuggestions/Row.java
package org.octo.seatingplacessuggestions.domain.seatingplacesuggestions;

import org.octo.seatingplacessuggestions.domain.seatingplacesuggestions.deepmodel.SeatingPlaceWithDistance;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import static org.octo.seatingplacessuggestions.domain.seatingplacesuggestions.deepmodel.adjacentseatingplace.AdjacentSeatingPlaces.offerAdjacentSeatingPlace;
import static org.octo.seatingplacessuggestions.domain.seatingplacesuggestions.deepmodel.middleoftherow.TheMiddleOfTheRow.offerSeatsNearerTheMiddleOfTheRow;

public record Row(String name, List<SeatingPlace> seatingPlaces) {

    public SeatingOptionIsSuggested suggestSeatingOption(PartyRequested partyRequested, PricingCategory pricingCategory) {

        var seatAllocation = new SeatingOptionIsSuggested(partyRequested, pricingCategory);

        for (var seat : offerAdjacentSeatsNearerTheMiddleOfRow(partyRequested, pricingCategory)) {
            if (seat.isAvailable() && seat.matchCategory(pricingCategory)) {
                seatAllocation.addSeat(seat);

                if (seatAllocation.matchExpectation())
                    return seatAllocation;

            }
        }
        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public Row addSeatingPlace(SeatingPlace seatingPlace) {
        List<SeatingPlace> newSeatingPlaces = new ArrayList<>( seatingPlaces );
        newSeatingPlaces.add(seatingPlace);
        return new Row(name, newSeatingPlaces);
    }

    public Row allocate(SeatingPlace seatingPlaceSuggested) {
        ArrayList<SeatingPlace> newSeatingPlaces = new ArrayList<>();
        seatingPlaces.forEach(seatingPlace -> {
            if (seatingPlaceSuggested.isSameLocation(seatingPlace))
                newSeatingPlaces.add(seatingPlaceSuggested.allocate());
            else
                newSeatingPlaces.add(seatingPlace);
        });
        return new Row(name, newSeatingPlaces);
    }

    private List<SeatingPlace> offerAdjacentSeatsNearerTheMiddleOfRow(PartyRequested partyRequested, PricingCategory pricingCategory)
    {
        var seatingPlacesWithDistance = offerSeatsNearerTheMiddleOfTheRow(this);

        if (partyRequested.partySize() > 1) {
            return offerAdjacentSeatingPlace(seatingPlacesWithDistance, pricingCategory, partyRequested.partySize());
        }

        return seatingPlacesWithDistance.stream()
                .sorted(Comparator.comparing(SeatingPlaceWithDistance::DistanceFromTheMiddleOfTheRow))
                .map(SeatingPlaceWithDistance::SeatingPlace)
                .filter(s -> s.matchCategory(pricingCategory))
                .filter(SeatingPlace::isAvailable)
                .limit(partyRequested.partySize())
                .toList();
    }
}
